File: Sources/Olaf/Network/OlafNetworkConfiguration.swift
import Foundation

/// Olaf network capture configuration.
public struct OlafNetworkConfiguration: Sendable {

    /// Also log request/response bodies. **On by default.** (Bodies are stored raw;
    /// set to `false` if not needed.)
    public var capturesBodies: Bool

    /// Also log request/response headers. **On by default** (headers are stored raw).
    public var capturesHeaders: Bool

    /// Maximum character count bodies are truncated to when logged.
    public var maxBodyLength: Int

    /// Upper limit (bytes) stored for **preview** of `image/*` response bodies. Images up to this
    /// size are appended to the record as base64 and shown on the detail screen; larger ones pass
    /// through with size info only (to avoid bloating RAM/disk). `0` → image preview disabled.
    public var maxImageBodyBytes: Int

    /// The category network records are logged under.
    public var category: LogCategory

    /// **Only** requests whose URL contains one of these parts (host/path) are captured. Empty = all.
    /// E.g. to capture only your own API: `["api-gateway", "myapp.com"]`.
    public var includedURLs: [String]

    /// Requests containing these URL parts are **skipped** (not captured — the request passes through as-is).
    /// E.g. to hide SDK noise: `["firebaseio", "crashlytics", "googleapis", "app-measurement"]`.
    public var excludedURLs: [String]

    /// Accept the server certificate unconditionally, **for the capture proxy only**. **`false` by default**
    /// (safe: system validation). Capture re-sends the request from its own URLSession; if the host trusts
    /// its own custom CA (e.g. an internal test gateway) or uses pinning, default system validation will
    /// reject that traffic (TLS -9807). If the host already trusts this traffic, set this to `true`
    /// (**non-prod only**) so capture accepts the certificates the host trusts. Olaf is already compiled
    /// under `#if !PROD`; live traffic is unaffected. NEVER enable in prod.
    public var allowsArbitraryServerTrustForCapture: Bool

    public init(
        capturesBodies: Bool = true,
        capturesHeaders: Bool = true,
        maxBodyLength: Int = 8000,
        maxImageBodyBytes: Int = 262_144,   // 256 KB
        category: LogCategory = .network,
        includedURLs: [String] = [],
        excludedURLs: [String] = [],
        allowsArbitraryServerTrustForCapture: Bool = false
    ) {
        self.capturesBodies = capturesBodies
        self.capturesHeaders = capturesHeaders
        self.maxBodyLength = max(0, maxBodyLength)
        self.maxImageBodyBytes = max(0, maxImageBodyBytes)
        self.category = category
        self.includedURLs = includedURLs.map { $0.lowercased() }
        self.excludedURLs = excludedURLs.map { $0.lowercased() }
        self.allowsArbitraryServerTrustForCapture = allowsArbitraryServerTrustForCapture
    }

    /// Should this URL be captured? (allow/deny filter — used in `OlafURLProtocol.canInit`)
    public func shouldCapture(_ url: URL?) -> Bool {
        guard let target = url?.absoluteString.lowercased() else { return includedURLs.isEmpty }
        if excludedURLs.contains(where: { target.contains($0) }) { return false }
        if !includedURLs.isEmpty { return includedURLs.contains(where: { target.contains($0) }) }
        return true
    }

    public static let `default` = OlafNetworkConfiguration()
}

File: Android/olaf/src/main/java/com/olaf/network/OlafNetworkConfiguration.kt
package com.olaf.network

import com.olaf.LogCategory

/**
 * Olaf network capture configuration.
 */
data class OlafNetworkConfiguration(

    /** Also log request/response bodies. **On by default** — bodies are stored raw. */
    val capturesBodies: Boolean = true,

    /** Also log request/response headers. **On by default** — headers are stored raw. */
    val capturesHeaders: Boolean = true,

    /** Bodies are truncated to this many characters when logged. */
    val maxBodyLength: Int = 8000,

    /**
     * Upper limit, in bytes, for storing an image response body as a **preview**. Images up to
     * this size are attached as base64 and rendered in the detail screen; larger ones only carry
     * their size, so RAM and disk don't balloon. `0` disables image previews.
     */
    val maxImageBodyBytes: Int = 262_144, // 256 KB

    /** The category network records are logged under. */
    val category: LogCategory = LogCategory.Network,

    /**
     * **Only** requests whose URL contains one of these fragments are captured; empty means all.
     * E.g. to capture just your own API: `listOf("api-gateway", "myapp.com")`.
     */
    val includedUrls: List<String> = emptyList(),

    /**
     * Requests containing one of these URL fragments are skipped entirely.
     * E.g. to hide SDK noise: `listOf("firebaseio", "crashlytics", "googleapis")`.
     */
    val excludedUrls: List<String> = emptyList(),

    /**
     * Decoders for bodies Olaf can't read as text — Protobuf and friends. Tried in order; the
     * first non-null result wins. See [BodyDecoder], which also covers compressed bodies.
     */
    val bodyDecoders: List<BodyDecoder> = emptyList()
) {

    private val normalizedIncluded = includedUrls.map { it.lowercase() }
    private val normalizedExcluded = excludedUrls.map { it.lowercase() }

    /** Allow/deny filter — exclusion always wins over inclusion. */
    fun shouldCapture(url: String?): Boolean {
        val target = url?.lowercase() ?: return normalizedIncluded.isEmpty()
        if (normalizedExcluded.any { target.contains(it) }) return false
        if (normalizedIncluded.isNotEmpty()) return normalizedIncluded.any { target.contains(it) }
        return true
    }

    companion object {
        val Default = OlafNetworkConfiguration()
    }
}
